### ISSUE TITLE
make it python3 ready

### ISSUE THREAD

**Issue by vpereira** [openSUSE/osc2#28]
The codebase is really clean. However it doesn't work with python3 :(

**Comment by lethliel** [comment]
I am working on it.

**Comment by bergmannf** [comment]
@lethliel I have branch at https://github.com/bergmannf/osc2/tree/python3, where i started working on also making the code `python2` and `3` compatible.

It stills fails for a lot of testcases under `python3`, but depending how much you already did, it maybe a simple starting point.

In `python2` all tests still pass with the updated imports at least.

**Comment by lethliel** [comment]
@bergmannf: Thanks. I will have a look at it. 

The biggest problem is that there is no python3 urlgrabber version. 

So I need to adapt urlgrabber to python3 (which is what I am trying to do atm) or to get 
rid of urlgrabber in the osc/osc2 code.

**Comment by vpereira** [comment]
The suggestion from upstream (urlgrabber) is to use python-pycurl...

**Comment by lethliel** [comment]
@vpereira: Yes. But we need to implement a few special things from urlgrabber, like MirrorGroup handling, which is not done by pycurl.

**Comment by Conan-Kudo** [comment]
@lethliel You could try your hand at porting `urlgrabber` to Python 3. It's on GitHub and accepts pull requests: https://github.com/rpm-software-management/urlgrabber

**Comment by lethliel** [comment]
@Conan-Kudo I am currently doing this. I am almost done. It works, but needs some more testing.

**Comment by Conan-Kudo** [comment]
@lethliel Awesome!

**Comment by Conan-Kudo** [comment]
@lethliel Any progress on this?

**Comment by lethliel** [comment]
I have ported urlgrabber to python3 and it works for us. But the test coverage of upstream urlgrabber is so low that I cannot be sure, that this will not break any other parts of the urlgrabber. 

(I learned that other distros have declared urlgrabber as unportable) 

I started to work on a substitution module for urlgrabber which provides the MirrorGroup functionality. 
This is IMHO the best way, because we do not rely on external modules on such a core function.

So still ongoing. And my main focus at the moment is fixing osc(1) to be python3 compatible. But as soon as osc(1) is done, fixing osc2 will be not that hard anymore.

**Comment by crazyscientist** [comment]
Hey guys, as some of you might know the SMASH project at SUSE Maintenance uses osc2 and we also would like to migrate our project to Python3, but are waiting for this issue.

If I can lend you a hand, just let me know ;)

**Comment by vpereira** [comment]
@crazyscientist, probably SMASH is the only project using it, AFAIK.. @lethliel were you able to finish the osc port to python3? Looking the TW package, we still using the version 2.7..

**Comment by Conan-Kudo** [comment]
Well, with all the warnings about how we shouldn't use it, I'm not surprised...

**Comment by crazyscientist** [comment]
Tell that to the people who chose to use osc2. As fas as I can tell osc2 is a better choice than osc, if one wants to use it as a module in another project.

**Comment by vpereira** [comment]
its for sure better designed, but has less features, I'm not sure they are keeping up with any API changes happening in the OBS, and it doesn't support python3. So is it a better choice? "Jein" :smiley: :smiley:

**Comment by marcus-h** [comment]
On 2018-09-12 05:33:18 -0700, Andreas Hasenkopf wrote:
> Hey guys, as some of you might know the SMASH project at SUSE Maintenance uses osc2 and we also would like to migrate our project to Python3, but are waiting for this issue.
> 
Hmm ok... since there seems to be some real interest in using osc2,
I'm going to start with the python3 port now. As a first step, I
just got rid of the urlgrabber dependency (see commit 2fdfd81 ("Get
rid of the urlgrabber dependency")).

**Comment by marcus-h** [comment]
On 2018-09-12 14:03:23 -0700, Victor Pereira wrote:
> its for sure better designed, but has less features, I'm not sure they are keeping up with any API changes happening in the OBS, and it doesn't support python3. So is it a better choice? "Jein" :smiley: :smiley:
> 
It should work with the recent OBS API. Btw., while we are talking
about the OBS API, this reminds me to resurrect + finish my
Hack Week 15 project... (in a nutshell: use the routes.rb
file to automatically generate a python OBS API (which will
be used by the osc2 code))

**Comment by crazyscientist** [comment]
Great, thanks for the quick commit. Maybe we should also merge the changes made by @lethliel and @bergmannf?

**Comment by vpereira** [comment]
>  (in a nutshell: use the routes.rb
> file to automatically generate a python OBS API (which will
> be used by the osc2 code))

That's a real good idea.

**Comment by crazyscientist** [comment]
I just merged the work of @bergmannf into commit 2fdfd81 and fixed all the issues arising from Python 3 distinguishing between byte and unicode strings (see commit 33616b8, which also contains my changes for pull request #32).

Testsuite is running successfully with Py2 and Py3.

Should I create a pull request?

**Comment by Conan-Kudo** [comment]
Yes, please do so.

**Comment by marcus-h** [comment]
First of all thanks for your efforts!

On 2018-09-13 12:00:37 +0000, Andreas Hasenkopf wrote:
> I just merged the work of @bergmannf into commit 2fdfd81 and fixed all the issues arising from Python 3 distinguishing between byte and unicode strings (see commit 33616b8, which also contains my changes for pull request #32).
> 
I just had a very very quick glance at the patch (commit 33616b8), but
this is exactly the kind of "porting" that I want to avoid: doing this
ad hoc encoding/decoding with the system's default encoding is quite
error prone/fragile.

Rationale:
- We heavily depend on the system's default encoding
  Suppose the system's default encoding is ascii and we pass a
  unicode string to osc2 => if we do an ad hoc encoding using the
  system's default encoding, we sooner or later will run into
  a UnicodeEncodeError...

  The same applies to ad hoc decoding.

- Bad API design
  We should clearly state what kind of data (string or bytes) a
  function expects, otherwise one might get unexpected results.
  For instance, assume the system's default encoding is utf-8,
  we pass a latin-1 string to osc2, osc2 ad hoc encodes the string
  using utf-8 and sends the encoded data to the server. If we
  now retrieve the data from the server we neither get back
  our original latin-1 string nor a necessarily valid latin-1
  encoding. This sucks:)


For these reasons, we will probably end up with an API where most
of the parameters are bytes (since the user of our API knows/should
know what kind of data she processes and how to encode/decode it...).
But I'm still thinking about this...

**Comment by crazyscientist** [comment]
Yes, that are all very valid arguments. As quite a few built-in functions (e.g. os.path.exists) in python3 can handle strings as well as bytes this adhoc en-/decoding would have been my choice.

Given the fact that pyhon2 allows the usage of Unicode characters in a `str` object, you will most likely move the problem just to python2 as developers might not have paid that much attention on encoding (and trust me, I have seen that more than enough). So basically osc2 would have to take care of en- or decoding of strings by trial and error as needed.

**Comment by crazyscientist** [comment]
@marcus-h , after talking to @lethliel I was wondering, if you are OK with our plan and whether you are working on the issue? If not, I would proceed as agreed with @lethliel ...
